### PR TITLE
use bind port 0 to solve port conflict in switch_unittest.cpp

### PR DESCRIPTION
**Pull request by huijunwu**
This PR is to replace #1833

The port returned by getFreePort() may be occupied before Start().
This PR let server bind port 0 and get the actual port after binding.

### CODE REVIEW

File: heron/common/src/cpp/network/baseserver.cpp [hunk 1]
```diff
@@ -106,6 +106,17 @@ sp_int32 BaseServer::Start_Base() {
     return -1;
   }
 
+  // fetch the port after bind/listen port 0
```
Comment (kramasamy) [inline heron/common/src/cpp/network/baseserver.cpp]: Can you add comments in the Start_Base method and also in options that new port is set when port 0 is provided.

File: heron/common/tests/cpp/network/switch_unittest.cpp [hunk 2]
```diff
@@ -26,13 +26,14 @@
...
 #include "threads/threads.h"
+#include "threads/spcountdownlatch.h"
 #include "network/network.h"
 #include "basics/modinit.h"
 #include "errors/modinit.h"
 #include "threads/modinit.h"
 #include "network/modinit.h"
 
-static const sp_uint32 SERVER_PORT = 61000;
+static sp_uint32 SERVER_PORT;
```
Comment (kramasamy) [inline heron/common/tests/cpp/network/switch_unittest.cpp]: Do we need to need to have a global variable for saving the port? Can we add something in the class itself?

File: heron/common/src/cpp/network/baseserver.cpp (hunk 1 above)
Comment (huijunw) [inline heron/common/src/cpp/network/baseserver.cpp]: comments added in the header.

File: heron/common/tests/cpp/network/switch_unittest.cpp (hunk 2 above)
Comment (huijunw) [inline heron/common/tests/cpp/network/switch_unittest.cpp]: removed global variable

File: heron/common/src/cpp/network/baseserver.cpp (hunk 1 above)
Comment (huijunw) [inline heron/common/src/cpp/network/baseserver.cpp]: comments added in the header.

**Comment by objmagic** [comment]
ping @kramasamy. Any more comments?

File: heron/common/src/cpp/network/baseserver.cpp [hunk 3]
```diff
@@ -106,6 +106,17 @@ sp_int32 BaseServer::Start_Base() {
...
+  if (AF_INET == options_.get_sin_family() && 0 == options_.get_port()) {
+    if (getsockname(listen_fd_, serv_addr, &sockaddr_len) != 0)  {
```
Comment (kramasamy) [inline heron/common/src/cpp/network/baseserver.cpp]: This code is being repeated in the other PRs as well. Can you please abstract it to a function in sock utils.cpp and call that function here.

Comment (huijunwu) [inline heron/common/src/cpp/network/baseserver.cpp]: This code appears only once.
This PR is the prerequisite of the other PRs. If this PR is merged, this code will not appear in the other PRs.

File: heron/common/tests/cpp/network/switch_unittest.cpp [hunk 4]
```diff
@@ -108,15 +110,20 @@ void terminate_server(sp_uint32 port) {
 }
 
 void start_test(sp_int32 nclients, sp_uint64 requests) {
+  sp_uint32 server_port = 0;
+  CountDownLatch* latch = new CountDownLatch(1);
+
   // start the server thread
-  std::thread sthread(start_server, SERVER_PORT);
+  std::thread sthread(start_server, &server_port, latch);
```
Comment (huijunwu) [inline heron/common/tests/cpp/network/switch_unittest.cpp]: The latch solves the race condition.
The server thread updates server_port, then latch is released, then the client thread can read server_port.

File: heron/common/tests/cpp/network/switch_unittest.cpp [hunk 5]
```diff
@@ -68,16 +67,19 @@ class Terminate : public Client {
 
 static TestServer* server_;
 
-void start_server(sp_uint32 port) {
+void start_server(sp_uint32* port, CountDownLatch* latch) {
```
Comment (huijunwu) [inline heron/common/tests/cpp/network/switch_unittest.cpp]: This function is run as thread with no return value.
I tried reference, but compiler complained. I suggest pointer.

Comment (kramasamy) [inline heron/common/tests/cpp/network/switch_unittest.cpp]: weird error though - let us stick with pointer

File: heron/common/tests/cpp/network/switch_unittest.cpp (hunk 4 above)
Comment (kramasamy) [inline heron/common/tests/cpp/network/switch_unittest.cpp]: Ok got it.

File: heron/common/src/cpp/network/baseserver.cpp (hunk 3 above)
Comment (huijunw) [inline heron/common/src/cpp/network/baseserver.cpp]: It is better not to put this piece of code in a separate function:
1. This code is not re-used in the other places.
2. Making it separate breaks the same style in the Start_Base()
3. Many variables involve in this code. It is not good tradeoff to pass many variables to a simple separate function.
4. If this code was in a separate function, why would not the other parts of the same Start_Base() be refactored into separate functions?

**Comment by objmagic** [comment]
@huijunwu LGTM. Feel free to merge this.

**Comment by huijunw** [comment]
Thanks @objmagic

**Pull request merged**